pkg/kv-cache/kv-block/in-memory.go

```diff
@@ -0,0 +1,200 @@
+
+	podsPerKey := make(map[Key][]string)
+	highestHitIdx := 0
+
+	for idx, key := range keys {
+		if pods, found := m.data.Get(key); found { //nolint:nestif // TODO: can this be optimized?
+			if pods == nil || pods.cache.Len() == 0 {
+				traceLogger.Info("no pods found for key, cutting search", "key", key.String())
+				return keys[:idx], podsPerKey, nil // early stop since prefix-chain breaks here
+			}
+
+			highestHitIdx = idx
+
+			if podIdentifierSet.Len() == 0 {
+				// If no pod identifiers are provided, return all pods
+				podsPerKey[key] = append(podsPerKey[key],
+					utils.SliceMap(pods.cache.Keys(), func(pod PodEntry) string {
+						return pod.PodIdentifier
+					})...)
+			} else {
+				// Filter pods based on the provided pod identifiers
+				for _, pod := range pods.cache.Keys() {
+					if podIdentifierSet.Has(pod.PodIdentifier) {
+						podsPerKey[key] = append(podsPerKey[key], pod.PodIdentifier)
+					}
+				}
+			}
+		} else {
+			traceLogger.Info("key not found in index", "key", key.String())
+		}
+	}
+
+	traceLogger.Info("lookup completed", "highest-hit-index", highestHitIdx,
+		"pods-per-key", podsPerKey)
+	return keys[:highestHitIdx+1], podsPerKey, nil
+}
+
+// Add adds a set of keys and their associated pod entries to the index backend.
+func (m *InMemoryIndex) Add(ctx context.Context, keys []Key, entries []PodEntry) error {
+	if len(keys) == 0 || len(entries) == 0 {
+		return fmt.Errorf("no keys or entries provided for adding to index")
+	}
+
+	traceLogger := klog.FromContext(ctx).V(5).WithName("kvblock.InMemoryIndex.Add")
+
+	for _, key := range keys {
+		podCache, found := m.data.Get(key) // bumps LRU timestamp if found
+		if !found {
+			cache, err := lru.New[PodEntry, time.Time](m.podCacheSize)
+			if err != nil {
+				return fmt.Errorf("failed to create pod cache for key %s: %w", key.String(), err)
+			}
+
+			podCache = &PodCache{
+				cache: cache,
+			}
+		}
+
+		for _, entry := range entries {
+			podCache.cache.Add(entry, time.Now()) // TODO: can this be batched to avoid multiple locks?
+		}
+
+		if !found {
+			m.data.Add(key, podCache) // TODO: conflicts here can lead to lost updates, fix locking
+		}
+
+		traceLogger.Info("added pods to key", "key", key.String(), "pods", entries)
+	}
+
+	return nil
+}
+
+// Evict removes a key and its associated pod entries from the index backend.
+func (m *InMemoryIndex) Evict(ctx context.Context, key Key, entries []PodEntry) error {
+	if len(entries) == 0 {
+		return fmt.Errorf("no entries provided for eviction from index")
+	}
+
+	traceLogger := klog.FromContext(ctx).V(5).WithName("kvblock.InMemoryIndex.Evict")
+
+	podCache, found := m.data.Get(key)
+	if !found || podCache == nil {
+		traceLogger.Info("key not found in index, nothing to evict", "key", key.String())
+		return nil
+	}
+
+	for _, entry := range entries {
+		podCache.cache.Remove(entry) // TODO: can this be batched to avoid multiple locks?
```

Comment:
same as on add. If you think batching is critical, pass the entire collection. 
It is relatively easy to send a collection of one if handcrafting tests